{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import time\n",
    "from datetime import datetime\n",
    "\n",
    "import metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build GIANA class\n",
    "class GIANA:\n",
    "    def __init__(self, input_data: pd.DataFrame) -> None:\n",
    "        self._input_data = input_data\n",
    "        self._processed_data = None\n",
    "        self._start_dir = os.getcwd()\n",
    "        self._giana_path = os.path.join(self._start_dir, 'functions/GIANA/')\n",
    "        self._save_path = os.path.join(self._start_dir, 'results')\n",
    "        self._cluster_map = None\n",
    "        self._settings = {\n",
    "            'Datetime': datetime.now().strftime('%Y%m%d_%H%M%S'),\n",
    "            'Model': 'GIANA',\n",
    "            }\n",
    "\n",
    "    def preprocess_data(self, species: str, antigen_species: str, chain_selection: str, min_vdj_score: int) -> None:\n",
    "        data = self._input_data\n",
    "        chain = chain_selection.lower()\n",
    "        self._settings['Species'] = species\n",
    "        self._settings['Antigen_species'] = antigen_species\n",
    "        self._settings['Chain'] = chain\n",
    "        self._settings['Minimum_VDJ_score'] = min_vdj_score\n",
    "        gene = \"TR\" + chain[0].upper()\n",
    "        cdr3_col_name = \"cdr3.\" + chain\n",
    "        variable_col_name = \"v.\" + chain\n",
    "        out = {}\n",
    "\n",
    "        # filter dataframe\n",
    "        data = data[(data[\"species\"] == species) & (data[\"vdjdb.score\"] >= min_vdj_score) & (data[\"gene\"] == gene)]\n",
    "        if antigen_species is not None:\n",
    "            data = data[data[\"antigen.species\"] == species]\n",
    "\n",
    "        # drop duplicates and null values\n",
    "        data = data.drop_duplicates()\n",
    "        data = data.dropna()\n",
    "\n",
    "        # filter input_data to only show genes belonging to passed chain\n",
    "        data = data.loc[data[\"gene\"] == gene]\n",
    "\n",
    "        out[cdr3_col_name] = data[\"cdr3\"].values\n",
    "        out[variable_col_name] = data[\"v.segm\"].values\n",
    "        out[\"epitope\"] = data[\"antigen.epitope\"].values\n",
    "\n",
    "        self._processed_data = pd.DataFrame.from_dict(out)\n",
    "    \n",
    "    def get_bio(self, data: pd.DataFrame) -> pd.DataFrame:\n",
    "        print('Getting Bio-ID on V, CDR3')\n",
    "        df = data.copy()\n",
    "        cols = [f'{x}.{self._settings[\"Chain\"]}' for x in ['v','cdr3']]\n",
    "        df.loc[:,'bio']=['-'.join(x) for x in df[cols].values.tolist()]\n",
    "        return df\n",
    "    \n",
    "    def run_giana(self, data: pd.DataFrame) -> None:\n",
    "        self._start_dir = os.getcwd()\n",
    "        os.chdir(self._giana_path)\n",
    "        cols = [f'{x}.{self._settings[\"Chain\"]}' for x in ['cdr3', 'v']]\n",
    "        df = data.copy()\n",
    "        df = df[cols]\n",
    "        df.columns = ['CDR3', 'V']\n",
    "        df.to_csv('input.txt', index=False, header=False, sep='\\t')\n",
    "        print(f'Clustering {len(df)} sequences with GIANA.')\n",
    "\n",
    "        # start timing process\n",
    "        start_time = time.time()\n",
    "        # call GIANA clustering script\n",
    "        %run GIANA4.1.py -f input.txt -O input_clustered.txt -v True\n",
    "        end_time = time.time()\n",
    "        self._settings['time_to_run'] = end_time - start_time\n",
    "        print(f'Elapsed time: {self._settings[\"time_to_run\"]} seconds.')\n",
    "\n",
    "        with open(os.path.join(self._giana_path, 'input_clustered.txt'), 'r') as f:\n",
    "            clusters = f.read().splitlines()[3:]\n",
    "            clusters = pd.DataFrame([x.split('\\t') for x in clusters], columns=['CDR3', 'cluster', 'V'])\n",
    "\n",
    "        # use bioidentities to map from output to input sequences\n",
    "        clusters = self.get_bio(clusters.rename(columns={'CDR3': f'cdr3.{self._settings[\"Chain\"]}', 'V': f'v.{self._settings[\"Chain\"]}'}))\n",
    "        self._cluster_map = {bio: cluster for bio, cluster in clusters[['bio', 'cluster']].values.tolist()}\n",
    "\n",
    "        # reset working directory\n",
    "        os.chdir(self._start_dir)\n",
    "    \n",
    "    def record_performance(self, data: pd.DataFrame)-> None:\n",
    "        df = data.copy()\n",
    "        df.loc[:, 'cluster'] = df['bio'].map(self._cluster_map)\n",
    "        counts = [l for l, n in df['cluster'].value_counts().reset_index()[['cluster','count']].values.tolist() if n<=1]\n",
    "        df.loc[:, 'cluster'] = df.loc[:, 'cluster'].replace(counts, np.nan)\n",
    "\n",
    "        # save clusters\n",
    "        if not os.path.exists(self._save_path):\n",
    "            os.mkdir(self._save_path)\n",
    "\n",
    "        df.to_csv(f'{self._save_path}/GIANA_{self._settings[\"Chain\"]}_clusters.csv')\n",
    "\n",
    "        # record performance\n",
    "        self._settings['n_clusters'] = len(df['cluster'].dropna().unique())\n",
    "\n",
    "        clusterscores = pd.DataFrame()\n",
    "        statistics = pd.DataFrame()\n",
    "        c_scores, stats = metrics.score(df, self._settings)\n",
    "        clusterscores = pd.concat([clusterscores, c_scores])\n",
    "        statistics = pd.concat([statistics, stats])\n",
    "\n",
    "        clusterscores.to_csv(f'{self._save_path}/GIANA_total.csv')\n",
    "        statistics.to_csv(f'{self._save_path}/GIANA_stats.csv')\n",
    "\n",
    "        print('GIANA results saved')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('./data/vdjdb.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "giana = GIANA(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "giana.preprocess_data(\"HomoSapiens\", \"SARS-CoV-2\", \"beta\", 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting Bio-ID on V, CDR3\n"
     ]
    }
   ],
   "source": [
    "bio_data = giana.get_bio(giana._processed_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clustering 634 sequences with GIANA.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/immanuelfagboyegun/miniconda3/envs/dsmp_groupol1/lib/python3.9/site-packages/sklearn/manifold/_mds.py:298: FutureWarning: The default value of `normalized_stress` will change to `'auto'` in version 1.4. To suppress this warning, manually set the value of `normalized_stress`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GIANA: Geometric Isometry based ANtigen-specific tcr Alignment\n",
      "Ultrafast short peptide alignment exclusively designed for large-scale adaptome analysis\n",
      "\n",
      "Input columns:\n",
      "1. CDR3 amino acid sequence (Starting from C, ending with the first F/L in motif [FL]G.G)\n",
      "2. Variable gene name in Imgt format: TRBVXX-XX*XX\n",
      "3. Joining gene name (optional)\n",
      "4. Frequency (optional)\n",
      "5. Other information (optional)\n",
      "\n",
      "!!! ALL amino acid letters must be CAPITAL !!!\n",
      "\n",
      "\n",
      "Processing input.txt\n",
      "Total time elapsed: 0.072672\n",
      "Maximum memory usage: 518.160384 MB\n",
      "Elapsed time: 0.9703741073608398 seconds.\n",
      "Getting Bio-ID on V, CDR3\n"
     ]
    }
   ],
   "source": [
    "giana.run_giana(bio_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No predictions made for:  {'GYNSYSVSNSEKHIM', 'FQDYIKSYL', 'YLSFYTAEQL', 'RMFPNAPYL', 'LLIRWQHFL', 'ESITGSLGPLL', 'LGYGFVNYI', 'GLLDEDFYA', 'GQVELGGGNAVEVCK', 'MEVDPIGHLY', 'SYMIMEIE', 'DPPALASTNAEVT', 'FRNEGIHL', 'MTEYKLVVVGARGVGKSALTIQLI', 'AVGVGKSAL', 'KASEKIFYV', 'FGDHPGHSY', 'RLLQCTQQAV', 'ENPVVHFFKNIVTP', 'AALPILFQV', 'TEDEHFEFY', 'MTEYKLVVVGAGDVGKSALTIQLI', 'GQVELGGGNAVEVCKGS', 'SLSKILDTV', 'VVMSWAPPV', 'VHFFKNIVTPRTPG', 'EVLPFFLFF', 'RLPGVLPRA', 'RVSTLRVSL', 'GAVGVGKSAL', 'MLIGIPVYV', 'RHDLPPYRVYL', 'YLSNIIPAL', 'KLTESLHKV', 'RLARLALVL', 'NHVVDISKSGLITIA', 'FGDVGSTLF', 'HYNYMCNSSCMGSMN', 'RLGEVRHPV', 'VYFFLPDHL', 'FLIYLDVSV', 'FYGKTILWF', 'SSCMGGMNWR', 'MVNTVAGAMK', 'MLFSHGLVK', 'MPYGYVLNEF', 'VLNGTVHPV', 'TLYSLTLLY', 'LLQCTQQAV', 'ALYGFVPVL', 'WMRLLPLL', 'MTLHGFMMY', 'TLREIRRYQK', 'SCMGGMNWRPILTII', 'CISSCNPNL', 'EEAAGIGIL', 'KALARALKEGRIR', 'NSKEETGHLENGN', 'ALLLQLFTL', 'VVRHCPHHERCSDSD', 'YLAMPFATPMEAELARRSLA', 'RILLVAASY', 'TYDTVHRHL', 'KIFGSLAFL', 'QAFWIDLFETIG', 'GDFGLATEKSRWSGS', 'EHEGSGPEL', 'GEEDGAGGHSL', 'VVVGAVGVGKSALTIQLIQN', 'YLYDRLLRV', 'FIASNGVKLV', 'VVVGADGVGK', 'HQNPVTGLLL', 'YVMAYVMAGVGS', 'LQPLALEGSLQKRG', 'APARLERRHSA', 'EMLFSHGLVK', 'ATGFKQSSKALQRPVAS', 'ALHGGWTTK', 'FLDEFMEGV', 'LYPEFIASI', 'ELLVRINRL', 'QPLALEGSLQKRG', 'WSYDSTLLAY', 'YLEPGPVTA', 'FSWGAEGQRPGF', 'NPVVHFFKNIVTPR', 'LAMPFATPM', 'TLWCSPIKV', 'WLIRETQPITK', 'RNTFRHSVVVPCE', 'GVYDGREHTV', 'QLCDVMFYL', 'VMAPRTLVL', 'EFTVSGNIL', 'RVRFFFPSL', 'LLLGIGILV', 'VLLGVKLFGV', 'YLSHLPLTCKF', 'VVGAVGVGK', 'HRRGSRSYV', 'ALLPGLPAA', 'AVGSYVYSV', 'MLYQHLLPL', 'MTEYKLVVVGAVGVGKSALTIQLI', 'QTNPVTLQY', 'ILCETCLIV', 'KLSHQLVLL', 'RYGSFSVTL', 'FMPDFDLHL'}\n",
      "GIANA results saved\n"
     ]
    }
   ],
   "source": [
    "giana.record_performance(bio_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3.beta</th>\n",
       "      <th>v.beta</th>\n",
       "      <th>epitope</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASSPQGLGTEAFF</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>ELAGIGILTV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASSLADRVNTEAFF</td>\n",
       "      <td>TRBV5-1*01</td>\n",
       "      <td>VHFFKNIVTPRTPG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CASSYVGNTGELFF</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>SLLMWITQV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CASSYVGNTGELFF</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CASSYVGNTGELFF</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>CASSLEVTYEQYF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>CASSLAGEGGNTIYF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>CASSEGQNYGYTF</td>\n",
       "      <td>TRBV27*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>CASGLDIHAFF</td>\n",
       "      <td>TRBV12-5*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>CASSPETGIGGQPQHF</td>\n",
       "      <td>TRBV7-2*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>634 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            cdr3.beta       v.beta         epitope\n",
       "0      CASSPQGLGTEAFF    TRBV28*01      ELAGIGILTV\n",
       "1     CASSLADRVNTEAFF   TRBV5-1*01  VHFFKNIVTPRTPG\n",
       "2      CASSYVGNTGELFF   TRBV6-5*01       SLLMWITQV\n",
       "3      CASSYVGNTGELFF   TRBV6-5*01       SLLMWITQC\n",
       "4      CASSYVGNTGELFF   TRBV6-5*01       SLLMWITQC\n",
       "..                ...          ...             ...\n",
       "629     CASSLEVTYEQYF   TRBV7-9*01      NLSALGIFST\n",
       "630   CASSLAGEGGNTIYF   TRBV7-9*01      NLSALGIFST\n",
       "631     CASSEGQNYGYTF    TRBV27*01      NLSALGIFST\n",
       "632       CASGLDIHAFF  TRBV12-5*01      NLSALGIFST\n",
       "633  CASSPETGIGGQPQHF   TRBV7-2*01      NLSALGIFST\n",
       "\n",
       "[634 rows x 3 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "giana._processed_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>antigen.species</th>\n",
       "      <th>reference.id</th>\n",
       "      <th>method</th>\n",
       "      <th>meta</th>\n",
       "      <th>cdr3fix</th>\n",
       "      <th>vdjdb.score</th>\n",
       "      <th>web.method</th>\n",
       "      <th>web.method.seq</th>\n",
       "      <th>web.cdr3fix.nc</th>\n",
       "      <th>web.cdr3fix.unmp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92766</th>\n",
       "      <td>30592</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSPGQGGDNEQFF</td>\n",
       "      <td>TRBV7-3*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-DQA1*05:01</td>\n",
       "      <td>HLA-DQB1*02:01</td>\n",
       "      <td>MHCII</td>\n",
       "      <td>PQQPFPQPEQPFP</td>\n",
       "      <td>...</td>\n",
       "      <td>Wheat</td>\n",
       "      <td>PMID:33927715</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSPGQGGDNEQFF\", \"cdr3_old\": \"CASSP...</td>\n",
       "      <td>0</td>\n",
       "      <td>sort</td>\n",
       "      <td>singlecell</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92767</th>\n",
       "      <td>30593</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAPQGATNKLIF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ32*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-DQA1*05:01</td>\n",
       "      <td>HLA-DQB1*02:01</td>\n",
       "      <td>MHCII</td>\n",
       "      <td>PQQPFPQPEQPFP</td>\n",
       "      <td>...</td>\n",
       "      <td>Wheat</td>\n",
       "      <td>PMID:33927715</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CAPQGATNKLIF\", \"cdr3_old\": \"CAPQGATN...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>singlecell</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92768</th>\n",
       "      <td>30593</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSLGAGGQETQYF</td>\n",
       "      <td>TRBV5-1*01</td>\n",
       "      <td>TRBJ2-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-DQA1*05:01</td>\n",
       "      <td>HLA-DQB1*02:01</td>\n",
       "      <td>MHCII</td>\n",
       "      <td>PQQPFPQPEQPFP</td>\n",
       "      <td>...</td>\n",
       "      <td>Wheat</td>\n",
       "      <td>PMID:33927715</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSLGAGGQETQYF\", \"cdr3_old\": \"CASSL...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>singlecell</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92769</th>\n",
       "      <td>30594</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CLVGGSGGYNKLIF</td>\n",
       "      <td>TRAV4*01</td>\n",
       "      <td>TRAJ4*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-DQA1*05:01</td>\n",
       "      <td>HLA-DQB1*02:01</td>\n",
       "      <td>MHCII</td>\n",
       "      <td>PQQPFPQPEQPFP</td>\n",
       "      <td>...</td>\n",
       "      <td>Wheat</td>\n",
       "      <td>PMID:33927715</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CLVGGSGGYNKLIF\", \"cdr3_old\": \"CLVGGS...</td>\n",
       "      <td>0</td>\n",
       "      <td>sort</td>\n",
       "      <td>singlecell</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92770</th>\n",
       "      <td>30594</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSSTAQETQYF</td>\n",
       "      <td>TRBV11-2*01</td>\n",
       "      <td>TRBJ2-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-DQA1*05:01</td>\n",
       "      <td>HLA-DQB1*02:01</td>\n",
       "      <td>MHCII</td>\n",
       "      <td>PQQPFPQPEQPFP</td>\n",
       "      <td>...</td>\n",
       "      <td>Wheat</td>\n",
       "      <td>PMID:33927715</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSSTAQETQYF\", \"cdr3_old\": \"CASSSTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>sort</td>\n",
       "      <td>singlecell</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92771 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       complex.id gene                  cdr3       v.segm      j.segm  \\\n",
       "0               1  TRA         CIVRAPGRADMRF  TRAV26-1*01   TRAJ43*01   \n",
       "1               1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "2               0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "3               2  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01   \n",
       "4               2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "...           ...  ...                   ...          ...         ...   \n",
       "92766       30592  TRB       CASSPGQGGDNEQFF   TRBV7-3*01  TRBJ2-1*01   \n",
       "92767       30593  TRA          CAPQGATNKLIF  TRAV12-2*01   TRAJ32*01   \n",
       "92768       30593  TRB       CASSLGAGGQETQYF   TRBV5-1*01  TRBJ2-5*01   \n",
       "92769       30594  TRA        CLVGGSGGYNKLIF     TRAV4*01    TRAJ4*01   \n",
       "92770       30594  TRB         CASSSTAQETQYF  TRBV11-2*01  TRBJ2-5*01   \n",
       "\n",
       "           species           mhc.a           mhc.b mhc.class antigen.epitope  \\\n",
       "0      HomoSapiens        HLA-B*08             B2M      MHCI        FLKEKGGL   \n",
       "1      HomoSapiens        HLA-B*08             B2M      MHCI        FLKEKGGL   \n",
       "2      HomoSapiens        HLA-B*08             B2M      MHCI        FLKEKGGL   \n",
       "3      HomoSapiens        HLA-B*08             B2M      MHCI        FLKEKGGL   \n",
       "4      HomoSapiens        HLA-B*08             B2M      MHCI        FLKEKGGL   \n",
       "...            ...             ...             ...       ...             ...   \n",
       "92766  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQQPFPQPEQPFP   \n",
       "92767  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQQPFPQPEQPFP   \n",
       "92768  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQQPFPQPEQPFP   \n",
       "92769  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQQPFPQPEQPFP   \n",
       "92770  HomoSapiens  HLA-DQA1*05:01  HLA-DQB1*02:01     MHCII   PQQPFPQPEQPFP   \n",
       "\n",
       "       ... antigen.species   reference.id  \\\n",
       "0      ...           HIV-1  PMID:15596521   \n",
       "1      ...           HIV-1  PMID:15596521   \n",
       "2      ...           HIV-1  PMID:15596521   \n",
       "3      ...           HIV-1  PMID:15596521   \n",
       "4      ...           HIV-1  PMID:15596521   \n",
       "...    ...             ...            ...   \n",
       "92766  ...           Wheat  PMID:33927715   \n",
       "92767  ...           Wheat  PMID:33927715   \n",
       "92768  ...           Wheat  PMID:33927715   \n",
       "92769  ...           Wheat  PMID:33927715   \n",
       "92770  ...           Wheat  PMID:33927715   \n",
       "\n",
       "                                                  method  \\\n",
       "0      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "1      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "2      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "3      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "4      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "...                                                  ...   \n",
       "92766  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "92767  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "92768  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "92769  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "92770  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "\n",
       "                                                    meta  \\\n",
       "0      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "1      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "2      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "3      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "4      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "...                                                  ...   \n",
       "92766  {\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...   \n",
       "92767  {\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...   \n",
       "92768  {\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...   \n",
       "92769  {\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...   \n",
       "92770  {\"cell.subset\": \"CD4+\", \"clone.id\": \"\", \"donor...   \n",
       "\n",
       "                                                 cdr3fix vdjdb.score  \\\n",
       "0      {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...           2   \n",
       "1      {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2   \n",
       "2      {\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...           2   \n",
       "3      {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...           2   \n",
       "4      {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2   \n",
       "...                                                  ...         ...   \n",
       "92766  {\"cdr3\": \"CASSPGQGGDNEQFF\", \"cdr3_old\": \"CASSP...           0   \n",
       "92767  {\"cdr3\": \"CAPQGATNKLIF\", \"cdr3_old\": \"CAPQGATN...           2   \n",
       "92768  {\"cdr3\": \"CASSLGAGGQETQYF\", \"cdr3_old\": \"CASSL...           2   \n",
       "92769  {\"cdr3\": \"CLVGGSGGYNKLIF\", \"cdr3_old\": \"CLVGGS...           0   \n",
       "92770  {\"cdr3\": \"CASSSTAQETQYF\", \"cdr3_old\": \"CASSSTA...           0   \n",
       "\n",
       "       web.method web.method.seq web.cdr3fix.nc web.cdr3fix.unmp  \n",
       "0            sort         sanger             no               no  \n",
       "1            sort         sanger             no               no  \n",
       "2            sort         sanger             no               no  \n",
       "3            sort         sanger             no               no  \n",
       "4            sort         sanger             no               no  \n",
       "...           ...            ...            ...              ...  \n",
       "92766        sort     singlecell             no               no  \n",
       "92767        sort     singlecell             no               no  \n",
       "92768        sort     singlecell             no               no  \n",
       "92769        sort     singlecell             no               no  \n",
       "92770        sort     singlecell             no               no  \n",
       "\n",
       "[92771 rows x 21 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "giana._input_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3.beta</th>\n",
       "      <th>v.beta</th>\n",
       "      <th>epitope</th>\n",
       "      <th>bio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASSPQGLGTEAFF</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>ELAGIGILTV</td>\n",
       "      <td>TRBV28*01-CASSPQGLGTEAFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASSLADRVNTEAFF</td>\n",
       "      <td>TRBV5-1*01</td>\n",
       "      <td>VHFFKNIVTPRTPG</td>\n",
       "      <td>TRBV5-1*01-CASSLADRVNTEAFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CASSYVGNTGELFF</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>SLLMWITQV</td>\n",
       "      <td>TRBV6-5*01-CASSYVGNTGELFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CASSYVGNTGELFF</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "      <td>TRBV6-5*01-CASSYVGNTGELFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CASSYVGNTGELFF</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "      <td>TRBV6-5*01-CASSYVGNTGELFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>CASSLEVTYEQYF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "      <td>TRBV7-9*01-CASSLEVTYEQYF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>CASSLAGEGGNTIYF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "      <td>TRBV7-9*01-CASSLAGEGGNTIYF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>CASSEGQNYGYTF</td>\n",
       "      <td>TRBV27*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "      <td>TRBV27*01-CASSEGQNYGYTF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>CASGLDIHAFF</td>\n",
       "      <td>TRBV12-5*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "      <td>TRBV12-5*01-CASGLDIHAFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>CASSPETGIGGQPQHF</td>\n",
       "      <td>TRBV7-2*01</td>\n",
       "      <td>NLSALGIFST</td>\n",
       "      <td>TRBV7-2*01-CASSPETGIGGQPQHF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>634 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            cdr3.beta       v.beta         epitope  \\\n",
       "0      CASSPQGLGTEAFF    TRBV28*01      ELAGIGILTV   \n",
       "1     CASSLADRVNTEAFF   TRBV5-1*01  VHFFKNIVTPRTPG   \n",
       "2      CASSYVGNTGELFF   TRBV6-5*01       SLLMWITQV   \n",
       "3      CASSYVGNTGELFF   TRBV6-5*01       SLLMWITQC   \n",
       "4      CASSYVGNTGELFF   TRBV6-5*01       SLLMWITQC   \n",
       "..                ...          ...             ...   \n",
       "629     CASSLEVTYEQYF   TRBV7-9*01      NLSALGIFST   \n",
       "630   CASSLAGEGGNTIYF   TRBV7-9*01      NLSALGIFST   \n",
       "631     CASSEGQNYGYTF    TRBV27*01      NLSALGIFST   \n",
       "632       CASGLDIHAFF  TRBV12-5*01      NLSALGIFST   \n",
       "633  CASSPETGIGGQPQHF   TRBV7-2*01      NLSALGIFST   \n",
       "\n",
       "                             bio  \n",
       "0       TRBV28*01-CASSPQGLGTEAFF  \n",
       "1     TRBV5-1*01-CASSLADRVNTEAFF  \n",
       "2      TRBV6-5*01-CASSYVGNTGELFF  \n",
       "3      TRBV6-5*01-CASSYVGNTGELFF  \n",
       "4      TRBV6-5*01-CASSYVGNTGELFF  \n",
       "..                           ...  \n",
       "629     TRBV7-9*01-CASSLEVTYEQYF  \n",
       "630   TRBV7-9*01-CASSLAGEGGNTIYF  \n",
       "631      TRBV27*01-CASSEGQNYGYTF  \n",
       "632      TRBV12-5*01-CASGLDIHAFF  \n",
       "633  TRBV7-2*01-CASSPETGIGGQPQHF  \n",
       "\n",
       "[634 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bio_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dsmp_groupol1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
